### PR TITLE
Support windows

### LINKED ISSUE #108: Unable to build with current nightly (rustc 1.62.0-nightly (311e2683e 2022-04-18))
Just fyi:
`cargo install hoard-rs` leads to:

image

### PR DESCRIPTION
**Pull request by Aziks0**
Closes #108 

[Crossterm](https://github.com/crossterm-rs/crossterm) is a cross-platform equivalent of `termion`.

The ui replacement is fairly easy because [tui](https://github.com/fdehau/tui-rs) supports, by default, `crossterm` as a backend.

The events are another thing tho. `crossterm` has no `Ctrl` modifier in its [key enum](https://docs.rs/crossterm/latest/crossterm/event/enum.KeyCode.html), so we need to use the module [event](https://docs.rs/crossterm/latest/crossterm/event/index.html) with `poll` and `read` which might require a bit more of refactoring. Not sure tho.

The changes are tested using Ubuntu 20.04 with WSL2.
_(I cannot test on Windows until I have completely removed `termion` from the dependencies)_

---

TODO: 
- [x] Raw mode
- [x] Alternate screen
- [x] Key events

### CODE REVIEW

**Comment by Aziks0** [comment]
I have removed the custom `event` module because we're forced to use `crossterm` own event to be able to handle key inputs.

`Crossterm` has a `poll` function for events. The description of the function is, from [crossterm doc](https://docs.rs/crossterm/latest/crossterm/event/index.html#examples):

> `poll()` waits for an `Event` for a given time period

It cannot be used by another thread than the main one, so I set the _"given time period"_ to 0 to avoid blocking the main thread. It should be noted that it means checking for key events every loop, no more 200ms tickrate.

I tested the changes on Ubuntu 20.04 with WSL2 and on Windows 10, it works as expected on both.

File: src/gui/list_search/controls.rs [hunk 1]
```diff
@@ -85,13 +133,21 @@ pub fn key_handler(
...
+        } => {
             state.input.pop();
             apply_filter(state, namespace_tabs, trove_commands);
             None
         }
-        Key::Char(c) => {
-            state.input.push(c);
+        // c
+        KeyEvent {
+            code: KeyCode::Char('c'),
```
Comment (Hyde46) [inline src/gui/list_search/controls.rs:147]: This event should handle any character key press to refine the search filter of all commands, not just the single character 'c'
```
 KeyEvent {
            code: KeyCode::Char(c),
            modifiers: KeyModifiers::NONE,
        } => {
            state.input.push(c);
            apply_filter(state, namespace_tabs, trove_commands);
            None
        }
```

File: src/gui/parameter_input/controls.rs [hunk 2]
```diff
@@ -22,12 +34,20 @@ pub fn key_handler(input: Key, app: &mut State) -> Option<HoardCommand> {
...
+            modifiers: KeyModifiers::NONE,
+        } => {
             app.input.pop();
             None
         }
-        Key::Char(c) => {
-            app.input.push(c);
+        // c
+        KeyEvent {
+            code: KeyCode::Char('c'),
```
Comment (Hyde46) [inline src/gui/parameter_input/controls.rs:47]: Same as in `list_search/control.rs`
```
 KeyEvent {
            code: KeyCode::Char(c),
            modifiers: KeyModifiers::NONE,
        } => {
            app.input.push(c);
            None
        }
```

**Review by Hyde46 — CHANGES_REQUESTED** [review]
Hey @Aziks0 , thanks a lot for your contribution, the refactoring looks great! :rocket: 
I've got 2 small requests for changes, then I think it'd be ready to go.

(Have you tried running it on a windows system?)

**Comment by Hyde46** [comment]
Just saw your comment that you already tested it on Windows 10, great!

On a separate note, I'd like to hook you up with a couple of stickers _if you're interested_. Hit me up at [denis.heid68@googlemail.com](mailto:denis.heid68@googlemail.com) with your address and i'll send you some, wherever you're located at
stickers
.

File: src/gui/list_search/controls.rs (hunk 1 above)
Comment (Aziks0) [reply src/gui/list_search/controls.rs]: Oops. I really messed up this one 😅. Fixed by last commit.

**Review by Hyde46 — APPROVED** [review]

**Comment by Hyde46** [comment]
Awesome!
Testing on my windows machine this week, before merging :rocket:

**Comment by glima** [comment]
This is neat! src/shell/hoard.pwsh would be a cherry on the cake! :D

**Pull request merged**